File: domain/src/main/java/com/github/anrimian/musicplayer/domain/interactors/sleep_timer/SleepTimerInteractor.java
package com.github.anrimian.musicplayer.domain.interactors.sleep_timer;

import com.github.anrimian.musicplayer.domain.interactors.player.LibraryPlayerInteractor;
import com.github.anrimian.musicplayer.domain.repositories.SettingsRepository;

import java.util.concurrent.TimeUnit;

import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.disposables.Disposable;
import io.reactivex.rxjava3.subjects.BehaviorSubject;

//TODO check android 5 and tablets
public class SleepTimerInteractor {

    public static final long NO_TIMER = -1;

    private final LibraryPlayerInteractor libraryPlayerInteractor;
    private final SettingsRepository settingsRepository;

    private final BehaviorSubject<Long> timerCountDownSubject = BehaviorSubject.createDefault(NO_TIMER);
    private final BehaviorSubject<SleepTimerState> sleepTimerStateSubject = BehaviorSubject.createDefault(SleepTimerState.DISABLED);

    private Disposable timerDisposable;

    private long remainingMillis;

    public SleepTimerInteractor(LibraryPlayerInteractor libraryPlayerInteractor,
                                SettingsRepository settingsRepository) {
        this.libraryPlayerInteractor = libraryPlayerInteractor;
        this.settingsRepository = settingsRepository;
    }

    public void start() {
        startSleepTimer(settingsRepository.getSleepTimerTime());
        sleepTimerStateSubject.onNext(SleepTimerState.ENABLED);
    }

    public void stop() {
        pause();
        remainingMillis = 0L;
        timerCountDownSubject.onNext(NO_TIMER);
        sleepTimerStateSubject.onNext(SleepTimerState.DISABLED);
    }

    public void pause() {
        if (timerDisposable != null) {
            timerDisposable.dispose();
        }
        sleepTimerStateSubject.onNext(SleepTimerState.PAUSED);
    }

    public void resume() {
        startSleepTimer(remainingMillis);
        sleepTimerStateSubject.onNext(SleepTimerState.ENABLED);
    }

    public Observable<Long> getSleepTimerCountDownObservable() {
        return timerCountDownSubject;
    }

    public Observable<SleepTimerState> getSleepTimerStateObservable() {
        return sleepTimerStateSubject;
    }

    public void setPlayLastSong(boolean playLastSong) {
        settingsRepository.setSleepTimerPlayLastSong(playLastSong);
    }

    public void setSleepTimerTime(long millis) {
        if (sleepTimerStateSubject.getValue() == SleepTimerState.ENABLED) {
            return;
        }
        settingsRepository.setSleepTimerTime(millis);
    }

    public long getSleepTimerTime() {
        return settingsRepository.getSleepTimerTime();
    }

    private void startSleepTimer(long timeMillis) {
        remainingMillis = timeMillis;
        timerDisposable = Observable.interval( 1, TimeUnit.SECONDS)
                .map(seconds -> remainingMillis -= 1000)
                .doOnSubscribe(d -> timerCountDownSubject.onNext(remainingMillis))
                .doOnNext(timerCountDownSubject::onNext)
                .takeUntil(millis -> millis < 0)
                .doOnComplete(this::onTimerFinished)
                .subscribe();
    }

    private void onTimerFinished() {
        libraryPlayerInteractor.pause();
        timerCountDownSubject.onNext(NO_TIMER);
        sleepTimerStateSubject.onNext(SleepTimerState.DISABLED);
    }

}

File: domain/src/main/java/com/github/anrimian/musicplayer/domain/interactors/sleep_timer/SleepTimerInteractor.kt
package com.github.anrimian.musicplayer.domain.interactors.sleep_timer

import com.github.anrimian.musicplayer.domain.interactors.player.LibraryPlayerInteractor
import com.github.anrimian.musicplayer.domain.models.player.PlayerState
import com.github.anrimian.musicplayer.domain.repositories.SettingsRepository
import com.github.anrimian.musicplayer.domain.utils.coroutines.onSuccess
import com.github.anrimian.musicplayer.domain.utils.coroutines.tickerLongValueFlow
import io.reactivex.rxjava3.core.Observable
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.onStart
import kotlinx.coroutines.flow.takeWhile
import kotlinx.coroutines.rx3.asFlow
import kotlinx.coroutines.rx3.asObservable
import kotlin.time.Duration.Companion.seconds

const val NO_TIMER = -1L

class SleepTimerInteractor(
    private val libraryPlayerInteractor: LibraryPlayerInteractor,
    private val settingsRepository: SettingsRepository,
    private val appComputationScope: CoroutineScope,
) {

    private val timerCountDownFlow = MutableStateFlow(NO_TIMER)
    private val sleepTimerStateFlow = MutableStateFlow(SleepTimerState.DISABLED)

    private var timerJob: Job? = null
    private var remainingMillis: Long = 0

    fun start() {
        startSleepTimer(settingsRepository.sleepTimerTime)
        sleepTimerStateFlow.value = SleepTimerState.ENABLED
    }

    fun stop() {
        timerJob?.cancel()
        remainingMillis = 0L
        timerCountDownFlow.value = NO_TIMER
        sleepTimerStateFlow.value = SleepTimerState.DISABLED
    }

    fun pause() {
        timerJob?.cancel()
        sleepTimerStateFlow.value = SleepTimerState.PAUSED
    }

    fun resume() {
        startSleepTimer(remainingMillis)
        sleepTimerStateFlow.value = SleepTimerState.ENABLED
    }

    fun getSleepTimerCountDownObservable(): Observable<Long> = timerCountDownFlow.asObservable()

    fun getSleepTimerStateFlow(): Flow<SleepTimerState> = sleepTimerStateFlow

    fun setPlayLastSong(playLastSong: Boolean) {
        settingsRepository.isSleepTimerPlayLastSong = playLastSong
    }

    fun setSleepTimerTime(millis: Long) {
        if (sleepTimerStateFlow.value == SleepTimerState.ENABLED) {
            return
        }
        settingsRepository.sleepTimerTime = millis
    }

    fun getSleepTimerTime() = settingsRepository.sleepTimerTime

    private fun startSleepTimer(timeMillis: Long) {
        remainingMillis = timeMillis
        timerJob = libraryPlayerInteractor.getPlayerStateObservable().asFlow()
            .flatMapLatest { playerState ->
                if (playerState == PlayerState.PLAY) {
                    tickerLongValueFlow(remainingMillis, -1000L, 1.seconds, 1.seconds)
                        .onEach { millis -> remainingMillis = millis }
                } else {
                    flowOf(remainingMillis)
                }
            }
            .onStart { timerCountDownFlow.value = remainingMillis }
            .onEach { millis -> timerCountDownFlow.value = millis }
            .takeWhile { millis -> millis >= 0 }
            .onSuccess(::onTimerFinished)
            .launchIn(appComputationScope)
    }

    private fun onTimerFinished() {
        libraryPlayerInteractor.pause()
        timerCountDownFlow.value = NO_TIMER
        sleepTimerStateFlow.value = SleepTimerState.DISABLED
    }

}